spellcheck/platform/linux/spellcheck_linux.cpp

```diff
@@ -1,38 +1,150 @@
-	return true;
+	void removeWord(const QString &word) {
+		auto w = word.toStdString();
+		for (const auto &validator : _validators) {
+			validator->remove_from_session(w);
+			validator->remove(w);
+		}
+	}
+
+	static EnchantSpellChecker *instance() {
+		static EnchantSpellChecker capsule;
+		return &capsule;
+	}
+
+private:
+	EnchantSpellChecker() {
+		if (!enchant::loader::do_explicit_linking()) return;
+
+		std::set<std::string> langs;
+		_brokerHandle = std::make_unique<enchant::Broker>();
+		_brokerHandle->list_dicts([](
+				const char *language,
+				const char *provider,
+				const char *description,
+				const char *filename,
+				void *our_payload) {
+			static_cast<decltype(langs)*>(our_payload)->insert(language);
+		}, &langs);
+		typedef std::unique_ptr<enchant::Dict> P;
+		try {
+			std::string langTag = QLocale::system().name().toStdString();
+			_validators.push_back(P(_brokerHandle->request_dict(langTag)));
+			langs.erase(langTag);
+		} catch (const enchant::Exception &e) {
+			// no first dictionary found
+		}
+		for (const std::string &language : langs) {
+			_validators.push_back(P(_brokerHandle->request_dict(language)));
+		}
+	}
+
+	EnchantSpellChecker(const EnchantSpellChecker&) = delete;
+	EnchantSpellChecker& operator =(const EnchantSpellChecker&) = delete;
+
+	std::unique_ptr<enchant::Broker> _brokerHandle;
+	std::vector<std::unique_ptr<enchant::Dict>> _validators;
+};
+
+}  // anonymous namespace
+
+bool Platform::Spellchecker::IsAvailable() {
+	return EnchantSpellChecker::instance()->isAvailable();
 }
 
-void FillSuggestionList(
-	const QString &wrongWord,
-	std::vector<QString> *optionalSuggestions) {
+void Platform::Spellchecker::KnownLanguages(std::vector<QString> *langCodes) {
+	*langCodes = EnchantSpellChecker::instance()->knownLanguages();
 }
 
-void AddWord(const QString &word) {
+bool Platform::Spellchecker::CheckSpelling(const QString &wordToCheck) {
+	return EnchantSpellChecker::instance()->checkSpelling(wordToCheck);
 }
 
-void RemoveWord(const QString &word) {
+void Platform::Spellchecker::FillSuggestionList(
+		const QString &wrongWord,
+		std::vector<QString> *variants) {
+	*variants = EnchantSpellChecker::instance()->findSuggestions(wrongWord);
 }
 
-void IgnoreWord(const QString &word) {
+void Platform::Spellchecker::AddWord(const QString &word) {
+	EnchantSpellChecker::instance()->addWord(word);
 }
 
-bool IsWordInDictionary(const QString &wordToCheck) {
-	return false;
+void Platform::Spellchecker::RemoveWord(const QString &word) {
+	EnchantSpellChecker::instance()->removeWord(word);
 }
 
-void CheckSpellingText(
-	const QString &text,
-	MisspelledWords *misspelledWordRanges) {
+void Platform::Spellchecker::IgnoreWord(const QString &word) {
+	EnchantSpellChecker::instance()->ignoreWord(word);
 }
 
-} // namespace Platform::Spellchecker
+bool Platform::Spellchecker::IsWordInDictionary(const QString &wordToCheck) {
+	return IsAvailable() && CheckSpelling(wordToCheck);
+}
```

Comment:
This function should give an information not about correct words, but about learned words (`Add to Dictionary`).

Reply:
An answer of the function affects the "Remove from Dictionary" item in context menu. Words that were added by a dictionary author (and that originally were considered correct), can also be deleted, there is no restriction in that.

Reply:
That is not how it's supposed to work.
`Add to Dictionary` and `Remove from Dictionary` work only with the custom list of words added by the user.
The first implementation with `personal.dic` looked right. But there was a question about where exactly this file was located.

Reply:
The `personal.dic` file was in working directory of the program. In case of Telegram Desktop it would be located at `~/.share/TelegramDesktop/personal.dic`. I abandoned the personal dictionary because in that case the words added by a user wouldn't be available in other apps. Instead, the new words is added to the first dictionary that corresponds to UI language. An actual list of added words will be placed in the `~/.config/enchant/` directory.

Reply:
Do you have examples of programs that use the Enchant as the system spellchecker?
I'm afraid that even if we will use system dictionaries as the user dictionary (which is logically incorrect), it will still be ignored by some important programs, for example, by browsers. So I was wondering what kind of programs respect Enchant's dictionary?

Reply:
I did a little searching and found that Pluma (text editor for MATE), Evolution (mail client for GNOME), Pidgin (universal chats, through libgtkspell), and perhaps other apps share the same list of added words which is stored in Enchant config directory. It seems they use the `add` method of already loaded dictionary, and don't create extra pwl (personal word list). As regards modern browsers, they all have feet of clay, I tested Firefox, and it's turning out to use an internal list of exceptions in a profile folder (file `persdict.dat`).

Reply:
Okay. So these applications can remove any correct word from the system dictionary?

Reply:
No, they can't. At least I don't see such feature.

Reply:
I see two options.
1. The library should remember the words added to the dictionary and be able to remove only these words.
2. The library should be disable to remove words from the dictionary. That is how Windows system spellchecker works now. Windows has no API to check the learned words and I should to do this manually, so I temporarily disabled this feature.